Content.Client/Tabletop/TabletopSystem.cs

```diff
@@ -0,0 +1,303 @@
+{
+    [UsedImplicitly]
+    public class TabletopSystem : SharedTabletopSystem
+    {
+        [Dependency] private readonly IInputManager _inputManager = default!;
+        [Dependency] private readonly IUserInterfaceManager _uiManger = default!;
+        [Dependency] private readonly IPlayerManager _playerManager = default!;
+
+        // Time in seconds to wait until sending the location of a dragged entity to the server again
+        private const float Delay = 1f / 10; // 10 Hz
+
+        private float _timePassed; // Time passed since last update sent to the server.
+        private IEntity? _draggedEntity; // Entity being dragged
+        private ScalingViewport? _viewport; // Viewport currently being used
+        private SS14Window? _window; // Current open tabletop window (only allow one at a time)
+        private IEntity? _table; // The table entity of the currently open game session
+
+        public override void Initialize()
+        {
+            CommandBinds.Builder
+                        .Bind(EngineKeyFunctions.Use, new PointerInputCmdHandler(OnUse, false))
+                        .Register<TabletopSystem>();
+
+            SubscribeNetworkEvent<TabletopPlayEvent>(OnTabletopPlay);
+            SubscribeLocalEvent<TabletopDraggableComponent, ComponentHandleState>(HandleComponentState);
+        }
+
+        public override void Update(float frameTime)
+        {
+            // If there is no player entity, return
+            if (_playerManager.LocalPlayer is not { ControlledEntity: { } playerEntity }) return;
+
+            if (StunnedOrNoHands(playerEntity))
+            {
+                StopDragging();
+            }
+
+            if (!CanSeeTable(playerEntity, _table))
+            {
+                StopDragging();
+                _window?.Close();
+                return;
+            }
+
+            // If no entity is being dragged or no viewport is clicked, return
+            if (_draggedEntity == null || _viewport == null) return;
+
+            // Make sure the dragged entity has a draggable component
+            if (!_draggedEntity.TryGetComponent<TabletopDraggableComponent>(out var draggableComponent)) return;
+
+            // If the dragged entity has another dragging player, drop the item
+            // This should happen if the local player is dragging an item, and another player grabs it out of their hand
+            if (draggableComponent.DraggingPlayer != null &&
+                draggableComponent.DraggingPlayer != _playerManager.LocalPlayer?.Session.UserId)
+            {
+                StopDragging(false);
+                return;
+            }
+
+            // Map mouse position to EntityCoordinates
+            var coords = _viewport.ScreenToMap(_inputManager.MouseScreenPosition.Position);
+
+            // Clamp coordinates to viewport
+            var clampedCoords = ClampPositionToViewport(coords, _viewport);
+            if (clampedCoords.Equals(MapCoordinates.Nullspace)) return;
```

Comment:
Again doing the cheap thing after the (only slightly more expensive here) thing...

Reply:
That's because I made `ClampPositionToViewport()` return `MapCoordinates.Nullspace` when there is no eye, but I added a check for nullspace to `ClampPositionToViewport()` just in case.